orchestra/model/build.py

```diff
@@ -102,9 +98,20 @@ def binary_archive_filename(self):
         component_commit = self.component.commit() or "none"
         return f'{component_commit}_{self.component.recursive_hash}.tar.gz'
 
-    @staticmethod
-    def _compute_build_hash(serialized_build):
-        return hashlib.sha1(json.dumps(serialized_build, sort_keys=True).encode("utf-8")).hexdigest()
+    @property
+    @lru_cache()
```

Comment:
I'm quite skeptical about `lru_cache` and the lazy approach. It has proven to be prolematic in the past.

1. Do you have data about how expensive this operation is?
2. Are you sure there isn't a single place where it makes sense to initialize it? In configuration, maybe?

Also, shouldn't we assert that certain fields that are initialized late, have been initialized?

Reply:
## Re: Q1
Quick tests timing orchestra using `time orc <cmd>` on my machine:
- `orc components --hashes` goes from 0.40s to 0.45s
- `orc graph -s` goes from 0.45/0.50s to 1.10s
I'd say `lru_cache` improves performance enough to be worth it.

## Re: Q2
Before this refactoring the hashes were indeed initialized externally by Configuration when ~~instanciating~~ instantiating the components, separating the code which computes the hashes for builds and components this way has these advantages:

- the build hash is computed using the actual values in the Build object, not by serializing the parsed json directly
- Build and Component know how to compute their own hash and don't care about details about the other one
- the hash is computer transparently, without the need to set it externally

I did implement a mechanism to check that nobody requests a hash before a component is initialized, namely `Component._hash_state` which makes this progression: `NOT_READY` -> `SELF_HASH_READY` -> `RECURSIVE_HASH_READY`, and is used to prohibit accesses at the wrong times. Components start as `NOT_READY` and become `SELF_HASH_READY` when the constructor is finished (as the constructor invokes e.g. the `Build` constructor I wanted to be 100% sure that `Build` cannot request the `Component`'s `self_hash`).
The transition to `RECURSIVE_HASH_READY` is done when calling `resolve_dependencies`, which must be done by who's calling `Component`'s constructor after it has instantiated all the `Components` (`resolve_dependencies` resolves dependency names into the actual actions which are only available after all the components and builds are created)

Reply:
I'm not suggesting to move out this stuff.
I'm just suggesting to a `self.hash` field that it's initialized in the constructor, once and for all.

The comparison is not with respect to dropping `lru_cache`, but to just compute it unconditionally in the constructor.

Initially, I thought this was not possible due to the need to resolve dependencies, but AFAIU all the information consumed here is already available in the constructor. Is this right?

This is probably not the case for `Component.recursive_hash`. But in that case you could initialize the `self.recursive_hash` field right at the end of `Component.resolve_dependencies`.

Reply:
I removed the lazy computation, performance is equivalent to the lazy, non cached ones in the examples above (makes sense, since they required computing all hashes for all components) and I don't see a significant performance drop in cases where there would be no need to compute all the hashes (e.g. `orc graph -s gmp`).